ipc: ipc_service: Add support for multiple instances

carlocaione opened the pull request:
The IPC service code is currently assuming that only one IPC instance does exist and the user can use the IPC service API to interface with that singleton instance.
    
This is a huge limitation and this patch is trying to fix this assumption introducing three major changes to the IPC service API:
    
- All the IPC instances are now supposed to be instantiated as a struct device. A new test is introduced to be used as skeleton for all the other backends.
    
- ipc_service_register_backend() is now removed (because multiple backends are now supported at the same time).
    
- All the other ipc_service_*() functions are now taking a struct device pointer as parameter to specify on which instance the user is going to act and operate.
    
In this patch the documentation is also extended to better clarify the terminology used.

carlocaione commented:
@mbolivar can you take a look at this from the DT layering prospective? I'm still a bit confused about that and I'd like to avoid doing the same mistake I did with the reserved-memory stuff.

arnopo commented on include/ipc/ipc_service.h:
```diff
@@ -70,20 +142,25 @@ struct ipc_ept_cfg {
  */
-int ipc_service_register_endpoint(struct ipc_ept **ept, const struct ipc_ept_cfg *cfg);
+int ipc_service_register_endpoint(struct ipc_ept **ept, const struct ipc_ept_cfg *cfg,
+				  const char *target_instance);
```
Nitpicking: maybe it would be better to have a shorter parameter name for the indentation `instance` seems explicit enough.

carlocaione replied:
ACK

arnopo commented on include/ipc/ipc_service.h:
```diff
@@ -70,20 +142,25 @@ struct ipc_ept_cfg {
  *  @param cfg Endpoint configuration.
+ *  @param target_instance Name of the instance to register the endpoint onto.
+ *                         Can be NULL when the backend implements an internal
+ *                         logic for endpoint-instance allocation.
```
Also null for backends that manage mono instance.proposal: 
Can be NULL depending on  the backend implementation ( default endpoint, or internal  logic for endpoint-instance allocation).

carlocaione replied:
ACK. I made the several cases explicit in the comment.

arnopo commented on samples/subsys/ipc/ipc_service/remote/src/main.c:
```diff
@@ -87,7 +87,7 @@ void app_task_1(void *arg1, void *arg2, void *arg3)
 	};
 
-	status = ipc_service_register_endpoint(&ept_1, &ept_cfg);
+	status = ipc_service_register_endpoint(&ept_1, &ept_cfg, NULL);
```
To be compatible with different backends, `instance = null `should also means default backend .

carlocaione replied:
ACK

arnopo reviewed (commented):
Please fin few comments.  else LGTM

mbolivar-nordic reviewed (changes requested):
@carlocaione I wasn't asking for a way to get things from strings, I was asking for a way to get them from the nodes themselves, similar to how DEVICE_DT_GET takes a node identifier, not a label property. That was what we had discussed at dev-review.

hubertmis reviewed (approved):
The code of ipc_service backends gets more and more confusing. It uses `DT_DRV_COMPAT` and `DEVICE_DT_DEFINE` but it's not a device driver. I agree using `const struct device` as a handle to backend instance is convenient, because as @mbolivar-nordic once said, in Linux everything is a file, but in Zephyr everything is a `const struct device`.

I believe the source of confusion is the name of the struct `device`, even though it is used to express much more concepts than device drivers. But it would be obviously not in scope of this PR to rename `const struct device`. That's why I approve, and looking forward for some changes like an alias to `const struct device` to make it obvious it can represent an instance of any software module in the system, not only device drivers.

mbolivar-nordic reviewed (changes requested):
temporary -1 while I am reviewing (doing this now)

mbolivar-nordic reviewed (dismissed):
Thanks, I like where this is going. But it looks like we can go further still and make this a real device API. That's my main design-level comment here; the rest are minor things about proper usage.

carlocaione committed:
ipc: ipc_service: Add support for multiple instances

The IPC service code is currently assuming that only one IPC instance
does exist and the user can use the IPC service API to interface with
that singleton instance.

This is a huge limitation and this patch is trying to fix this
assumption introducing three major changes to the IPC service API:

- All the IPC instances are now supposed to be instantiated as a struct
  device. A new test is introduced to be used as skeleton for all the
  other backends.

- ipc_service_register_backend() is now removed (because multiple
  backends are now supported at the same time).

- All the other ipc_service_*() functions are now taking a struct device
  pointer as parameter to specify on which instance the user is going to
  act and operate.

In this patch the documentation is also extended to better clarify the
terminology used.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
ipc_service: mi: Convert to device driver

As part of the work to support multiple IPC instances / backends using
IPC service, the static vrings mi code must be reworked to resemble a
classic device driver.

Fix also the sample using it.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

hubertmis reviewed (changes requested):
NACKing until my comments are resolved

hubertmis reviewed (approved):
I like APIs and data flow between modules now :+1: 
The only remaining thing is `prio` in `struct ipc_ept_cfg`, which I understand will be removed later.

carlocaione commented on tests/subsys/ipc/ipc_service/src/backend.c:
```diff
@@ -0,0 +1,91 @@
+		const void *p_data, size_t len)
+{
+	const struct backend_config_t *config;
+	const struct device *instance;
```
yup

carlocaione commented on tests/subsys/ipc/ipc_service/src/backend.c:
```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2021-2022, Carlo Caione <ccaione@baylibre.com>
```
ahaah, I'm already planning for when this will be actually merged ;)

carlocaione commented:
@mbolivar-nordic gentle ping on this

mbolivar-nordic commented:
> stale

ack
